Initial implementation

JesseCrocker opened the pull request:
Still to do:
- [x] load a json file that specifies options for background styles
- [x] generate style rules for overlay data and merge into style

jczaplew commented on mapUtils.js:
```diff
@@ -0,0 +1,54 @@
+      return zoom;
+    }
+  }
+  return 15;
```
Why is the default zoom 15?

jczaplew commented on mapUtils.js:
```diff
@@ -0,0 +1,54 @@
+
+exports.addOverlayDataToStyle = function(style, overlay) {
+  style.sources["overlay"] = { type: "geojson", data: overlay };
+  style.layers.push(overlayFillLayerDef);
```
`push` takes an arbitrary number of arguments, so these can all be added to `style.layers` at once

jczaplew commented on package.json:
```diff
@@ -0,0 +1,24 @@
+  "dependencies": {
+    "@mapbox/mapbox-gl-native": "^4.1.0",
+    "@mapbox/sphericalmercator": "^1.1.0",
+    "@turf/turf": "^5.1.6",
```
Since `bbox` is the only `turf` function used we could install that package by itself `@turf/bbox`

jczaplew commented on mapUtils.js:
```diff
@@ -0,0 +1,54 @@
+const sm = new (require("@mapbox/sphericalmercator"))({ size: 512 });
+
+const overlayLineLayerDef = {
```
I wonder if these overlay layers should be in their own file or somewhere that makes it easier to configure 🤔

JesseCrocker commented on README.md:
```diff
@@ -1,2 +1,55 @@
+/**zoom**/**lon**/**lat**/**width**/**height**/**background**.**format**
+Example `/10.5/45/-111/200/200/default.png`
+Post data is optional.
+Zoom levels are based on 512 pixel tiles, so they may be 1 off from expected.
```
No, fractional zooms work fine.

JesseCrocker commented on getMap.js:
```diff
@@ -0,0 +1,137 @@
+  if (maxMapUses > 0) {
+    factory["validate"] = function(resource) {
+      debug("validate");
+      return new Promise(function(resolve) {
```
Because thats what the docs for generic-pool say to do
> validate: a function that the pool will call if it wants to validate a resource. It should accept one argument resource where resource is whatever factory.create made. Should return a Promise that resolves a boolean where true indicates the resource is still valid or false if the resource is invalid.

jczaplew commented on README.md:
```diff
@@ -1,2 +1,55 @@
+
+Request a map centered on a given coordinate, and a given zoom level.
+/**zoom**/**lon**/**lat**/**width**/**height**/**background**.**format**
+Example `/10.5/45/-111/200/200/default.png`
```
Lat and long are switched here 
```suggestion
Example `/10.5/-111/45/200/200/default.png`
```

JesseCrocker commented:
@jczaplew Can you take another look at this? I addressed most of your comments, but ignored a few of them because they work and i dont really anticipate touching this code much in the future.
